cpp/include/cudf/detail/iterator.cuh

```diff
@@ -62,12 +63,16 @@ namespace detail {
  * iter[n] == n * n
  * @endcode
  *
- * @param start The starting value of the counting iterator
+ * @param start The starting value of the counting iterator (must be size_type or smaller type).
  * @param f The unary function to apply to the counting iterator.
  * @return A transform iterator that applies `f` to a counting iterator
  */
-template <typename UnaryFunction>
-CUDF_HOST_DEVICE inline auto make_counting_transform_iterator(cudf::size_type start,
+template <typename CountingIterType,
+          typename UnaryFunction,
+          typename = cuda::std::enable_if_t<cuda::std::is_integral_v<CountingIterType> and
+                                            cuda::std::numeric_limits<CountingIterType>::max() <=
+                                              cuda::std::numeric_limits<cudf::size_type>::max()>>
```

Comment:
With SFINAE, if the code fails to compile, the compiler will give a very long, cryptic error message. If we use `static_assert`, the compiler will just direct us to the failed assert message, which is much cleaner. In the future I recommend to use `static_assert` instead.

Reply:
In addition, this comparison would generate compiler warning when comparing signed vs unsigned number, (if `CountingIterType` is unsigned).

Reply:
Imo, `enable_if` is usually used to remove one function overload in favor of another one. If our intention is solely to place a requirement on `CountingIterType` and generate a compile error when the requirement is not satisfied, then `static_assert(condition, error_msg)` is probably the best option in C++17 because of the advantage Nghia mentioned.

The sign-compare warning usually applies to non-`constexpr` if I'm not mistaken, so the existing solution should be fine. If we are strict about the sign though, the following can be considered:

```
static_cast<cuda::std::make_signed_t<CountingIterType>>(cuda::std::numeric_limits<CountingIterType>::max()) <=
                                              cuda::std::numeric_limits<cudf::size_type>::max()>
```
Actually, the following is probably a better alternative without the sign issue:

```
cuda::std::numeric_limits<CountingIterType>::digits <= cuda::std::numeric_limits<cudf::size_type>::digits
```

Reply:
Thanks for the idea with using `cuda::std::numeric_limits<CountingIterType>::digits`. I considered using `static_assert` in one of the commits but then reverted for some reason. Going to use a static_assert to make the compilation error more friendly.

Reply:
Oh I didn't know that. Thanks :D

Reply:
> Oh I didn't know that. Thanks :D

Me neither so I had to verify for myself too 😆 : https://godbolt.org/z/GqP5KrWKz